feat: Add plugin call variables to sidecar plugin discovery (#9273)

pierrecregut opened the pull request:
Gives access to variables declared in the call of the plugin in the application
manifest to the discover command run on the CMP server.

Variables are prefixed with ENV_ to avoid security issues (plugin call
overiding important variables).
Modified variable names are also given to the generate command but the variables
with unmodified names are also defined to preserve the old behavior until it is
deprecated.

Fixes #9273

Signed-off-by: Pierre Crégut <pierre.cregut@orange.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

crenshaw-dev reviewed (approved):
lgtm, thanks!

crenshaw-dev commented on reposerver/repository/repository.go:
```diff
@@ -1294,27 +1294,39 @@ func getPluginEnvs(envVars *v1alpha1.Env, q *apiclient.ManifestRequest, creds gi
+			if legacy {
+				env = append(env, fmt.Sprintf("%s=%s", entry.Name, newValue))
+			}
+			env = append(env, fmt.Sprintf("ENV_%s=%s", entry.Name, newValue))
```
@pierrecregut sorry for the 2nd review, but I'm wondering if you would change this to `ARGOCD_ENV_`. I just suddenly became worried that this might have collisions.

pierrecregut replied:
Done.

crenshaw-dev commented on reposerver/repository/repository.go:
```diff
@@ -1294,27 +1294,39 @@ func getPluginEnvs(envVars *v1alpha1.Env, q *apiclient.ManifestRequest, creds gi
-			pluginEnv[i].Value = parsedEnv.Envsubst(j.Value)
+		for _, entry := range pluginEnv {
+			newValue := parsedEnv.Envsubst(entry.Value)
+			if legacy {
```
@leoluz suggested we consider removing the legacy behavior completely in 2.4. I think it's worth considering moving aggressively, since it's a security enhancement. I've added a note to Thursday's contributor meeting to discuss. https://docs.google.com/document/d/1xkoFkVviB70YBzSEa4bDnu-rUZ1sIFtwKKG1Uw8XsY8/edit#

crenshaw-dev replied:
> But this will also break published plugins.

True. I think it's only justified if we think the security benefit is significant and we can communicate the change effectively.

> Have you considered building the env array in reverse order ? This way a user cannot override an existing variable.

I don't think that's sufficient to resolve the security concern. Tooling in the CMP might pick up env vars that aren't explicitly set elsewhere and allow malicious behavior. 

> I am not sure if this is really intended.

I think allowing the user to override CMP-side vars was intentional, but I think it was a bad idea.

By prefixing the env var names, I think we'll make it so that the array order doesn't matter - there should be no overlap in env var names.

pierrecregut replied:
The side note was about the fact that both the repo server and the cmp server appended their environment variables to the environment used in the call (priority for the repoServer. For example, the value of PATH is the wrong one). I think it is a side effect of the fact that `getPluginEnvs`  is also used for plugin loaded in the repoServer.

crenshaw-dev replied:
Oh wait. So the repo-server PATH var (and other vars) end up taking precedence over the sidecar? If so, yikes.

pierrecregut replied:
I have made a real test adding a variable TEST to both containers with value either repo-server or cmp-server and I can confirm what I inferred from the code TEST value was repo-server in the plugin. I can add a boolean variable remote to distinguish between the behavior for https://github.com/argoproj/argo-cd/blob/08009211ff209ea20ee69155cd80f9397545c8e0/reposerver/repository/repository.go#L1255 and the behavior required for sidecars.

crenshaw-dev replied:
Agreed, separate PR makes sense.

pierrecregut replied:
I have pushed #9397

crenshaw-dev commented:
Thanks for making the change! I'll let you know the outcome of tomorrow's meeting, if you're not able to attend yourself.

crenshaw-dev commented:
@pierrecregut the conclusion was that we should probably go ahead and aggressively remove the legacy behavior, but that 1) we should document the change and migration instructions in the upgrade guide and release notes (I'll open a PR) and 2) we should post in Slack about the change to make sure the community isn't strongly opposed (I'll do this as well).

For now if you want to go ahead and change this PR to remove legacy behavior, I'll approve. We can revert if there's strong opposition in Slack.

cc @todaywasawesome @leoluz @jopit

pierrecregut commented:
I have removed the implementation of the legacy behavior and implemented the "aggressive" approach. I have *removed* the small documentation change letting you implement it in a separate PR.

This PR is rebased on top of the bug fixing PR #9397

crenshaw-dev commented on reposerver/repository/repository.go:
```diff
@@ -1270,8 +1270,14 @@ func runConfigManagementPlugin(appPath, repoRoot string, envVars *v1alpha1.Env,
-func getPluginEnvs(envVars *v1alpha1.Env, q *apiclient.ManifestRequest, creds git.Creds) ([]string, error) {
-	env := append(os.Environ(), envVars.Environ()...)
+func getPluginEnvs(envVars *v1alpha1.Env, q *apiclient.ManifestRequest, creds git.Creds, remote bool) ([]string, error) {
+	env := envVars.Environ()
```
Note: this is a breaking change. Plugins that rely on the environment variables from the main API server container will need that env var duplicated.

crenshaw-dev commented on util/app/discovery/discovery.go:
```diff
@@ -33,7 +33,7 @@ func Discover(ctx context.Context, repoPath string, enableGenerateManifests map[
 
 	// Check if it is CMP
-	conn, _, err := DetectConfigManagementPlugin(ctx, repoPath)
+	conn, _, err := DetectConfigManagementPlugin(ctx, repoPath, []string{})
```
It's weird to me that this doesn't send any env. I think it's wrong. But I think that's a separate bug, not an issue with this PR.

crenshaw-dev reviewed (approved):
lgtm, thanks for your patience!
